### ISSUE TITLE
Values cannot be echoed or subscribed to, set_stream_rate service call inconsistent behavior

### ISSUE THREAD

**Issue by weicong96** [mavlink/mavros#1657]
This is only bug and feature tracker, please use it
to report bugs or request features.

----

### Issue details

I have read other issues filed that set_stream_rate is required to resolve the issue with no satellites found. However, calling set_stream_rate has inconsistent behavior for me.

When it works:
![image](https://user-images.githubusercontent.com/9016780/143509797-e93b3eaa-e9e3-49b7-ba50-d10c96cb88eb.png)

When it does not work it hangs indefinitely without any response, it has past 7 minutes at the time of writing this:
![image](https://user-images.githubusercontent.com/9016780/143509854-10d9a008-c131-4fc1-adad-3a3721a072d4.png)

I would be able to get telemetry data coming in if I had connected to QGC or Mission Planner, disconnect it and connect via mavros.

in the case of stream rate call hanging, only able to echo /mavros/state and not other topics.

Is there a solution for setting stream rate?

### MAVROS version and platform

Mavros: 2.0.3
ROS: Foxy
Ubuntu: 18.04

### Autopilot type and version

[ x ] ArduPilot
[  ] PX4

Version: 4.0.3

### Node logs

```
[INFO] [1637885590.807011997] [mavros_node]: Starting mavros_node container
[INFO] [1637885590.807160800] [mavros_node]: FCU URL: tcp://localhost:5760
[INFO] [1637885590.807176798] [mavros_node]: GCS URL: udp://@192.168.1.117
[INFO] [1637885590.807184537] [mavros_node]: UAS Prefix: /uas1
[INFO] [1637885590.807191724] [mavros_node]: Starting mavros router node
[INFO] [1637885591.014065182] [mavros_router]: Built-in SIMD instructions: SSE, SSE2
[INFO] [1637885591.014152440] [mavros_router]: Built-in MAVLink package version: 2021.10.10
[INFO] [1637885591.014179103] [mavros_router]: Known MAVLink dialects: common ardupilotmega ASLUAV all development icarous matrixpilot paparazzi standard uAvionix ualberta
[INFO] [1637885591.014190500] [mavros_router]: MAVROS Router started
[INFO] [1637885591.014252918] [mavros_router]: Requested to add endpoint: type: 0, url: tcp://localhost:5760
[INFO] [1637885591.014315798] [mavros_router]: Endpoint link[1000] created
[INFO] [1637885591.029710371] [mavros_router]: link[1000] opened successfully
[INFO] [1637885591.030049807] [mavros_router]: Requested to add endpoint: type: 1, url: udp://@192.168.1.117
[INFO] [1637885591.030137365] [mavros_router]: Endpoint link[1001] created
[INFO] [1637885591.030377034] [mavros_router]: link[1001] opened successfully
[INFO] [1637885591.030431283] [mavros_router]: Requested to add endpoint: type: 2, url: /uas1
[INFO] [1637885591.030457194] [mavros_router]: Endpoint link[1002] created
[INFO] [1637885591.031851626] [mavros_router]: link[1002] opened successfully
[INFO] [1637885591.031904712] [mavros_node]: Starting mavros uas node
[INFO] [1637885592.155947944] [mavros_router]: link[1000] detected remote address 1.1
[WARN] [1637885592.222138016] [mavros]: diagnostic_updater: No HW_ID was set. This is probably a bug. Please report it. For devices that do not have a HW_ID, set this value to 'none'. This warning only occurs once all diagnostics are OK. It is okay to wait until the device is open before calling setHardwareID.
[INFO] [1637885592.322537428] [mavros]: UAS Executor started
[INFO] [1637885593.056227582] [mavros]: Plugin actuator_control created
[INFO] [1637885593.056306540] [mavros]: Plugin actuator_control added to executor
[INFO] [1637885593.056342946] [mavros]: Plugin actuator_control initialized
[INFO] [1637885593.070531455] [mavros]: Plugin altitude created
[INFO] [1637885593.070601932] [mavros]: Plugin altitude added to executor
[INFO] [1637885593.070620646] [mavros]: Plugin altitude initialized
[INFO] [1637885593.081712058] [mavros]: Plugin command created
[INFO] [1637885593.081801842] [mavros]: Plugin command added to executor
[INFO] [1637885593.081831923] [mavros]: Plugin command initialized
[INFO] [1637885593.113814613] [mavros]: Plugin ftp created
[INFO] [1637885593.114010087] [mavros]: Plugin ftp added to executor
[INFO] [1637885593.114052618] [mavros]: Plugin ftp initialized
[INFO] [1637885593.124715964] [mavros]: Plugin geofence created
[INFO] [1637885593.124787514] [mavros]: Plugin geofence added to executor
[INFO] [1637885593.124806940] [mavros]: Plugin geofence initialized
[INFO] [1637885593.156204750] [mavros]: Plugin global_position created
[INFO] [1637885593.156268151] [mavros]: Plugin global_position added to executor
[INFO] [1637885593.156284450] [mavros]: Plugin global_position initialized
[INFO] [1637885593.165488369] [mavros]: Plugin home_position created
[INFO] [1637885593.165536674] [mavros]: Plugin home_position added to executor
[INFO] [1637885593.165550798] [mavros]: Plugin home_position initialized
[INFO] [1637885593.173338794] [mavros]: Plugin imu created
[INFO] [1637885593.173393414] [mavros]: Plugin imu added to executor
[INFO] [1637885593.173403568] [mavros]: Plugin imu initialized
[INFO] [1637885593.180919980] [mavros]: Plugin local_position created
[INFO] [1637885593.180966350] [mavros]: Plugin local_position added to executor
[INFO] [1637885593.180979010] [mavros]: Plugin local_position initialized
[INFO] [1637885593.186052410] [mavros]: Plugin manual_control created
[INFO] [1637885593.186091432] [mavros]: Plugin manual_control added to executor
[INFO] [1637885593.186161419] [mavros]: Plugin manual_control initialized
[INFO] [1637885593.198972054] [mavros]: Plugin param created
[INFO] [1637885593.199022263] [mavros]: Plugin param added to executor
[INFO] [1637885593.199138268] [mavros]: Plugin param initialized
[INFO] [1637885593.209673701] [mavros]: Plugin rallypoint created
[INFO] [1637885593.209749019] [mavros]: Plugin rallypoint added to executor
[INFO] [1637885593.209769839] [mavros]: Plugin rallypoint initialized
[INFO] [1637885593.216560978] [mavros]: Plugin rc_io created
[INFO] [1637885593.216604281] [mavros]: Plugin rc_io added to executor
[INFO] [1637885593.216614164] [mavros]: Plugin rc_io initialized
[INFO] [1637885593.222431791] [mavros]: Plugin setpoint_accel created
[INFO] [1637885593.222474713] [mavros]: Plugin setpoint_accel added to executor
[INFO] [1637885593.222494871] [mavros]: Plugin setpoint_accel initialized
[INFO] [1637885593.232645249] [mavros]: Plugin setpoint_attitude created
[INFO] [1637885593.232693002] [mavros]: Plugin setpoint_attitude added to executor
[INFO] [1637885593.232707436] [mavros]: Plugin setpoint_attitude initialized
[INFO] [1637885593.242006902] [mavros]: Plugin setpoint_position created
[INFO] [1637885593.242051990] [mavros]: Plugin setpoint_position added to executor
[INFO] [1637885593.242066765] [mavros]: Plugin setpoint_position initialized
[INFO] [1637885593.255791211] [mavros]: Plugin setpoint_raw created
[INFO] [1637885593.255842673] [mavros]: Plugin setpoint_raw added to executor
[INFO] [1637885593.255854661] [mavros]: Plugin setpoint_raw initialized
[INFO] [1637885593.273567821] [mavros]: Plugin setpoint_trajectory created
[INFO] [1637885593.273617078] [mavros]: Plugin setpoint_trajectory added to executor
[INFO] [1637885593.273870670] [mavros]: Plugin setpoint_trajectory initialized
[INFO] [1637885593.282570893] [mavros]: Plugin setpoint_velocity created
[INFO] [1637885593.282605536] [mavros]: Plugin setpoint_velocity added to executor
[INFO] [1637885593.282638464] [mavros]: Plugin setpoint_velocity initialized
[INFO] [1637885593.294327355] [mavros]: Plugin sys_status created
[INFO] [1637885593.294392660] [mavros]: Plugin sys_status added to executor
[INFO] [1637885593.294417268] [mavros]: Plugin sys_status initialized
[INFO] [1637885593.304528242] [mavros.time]: TM: Timesync mode: MAVLINK
[INFO] [1637885593.306600815] [mavros]: Plugin sys_time created
[INFO] [1637885593.306647315] [mavros]: Plugin sys_time added to executor
[INFO] [1637885593.306656407] [mavros]: Plugin sys_time initialized
[INFO] [1637885593.321633137] [mavros]: Plugin waypoint created
[INFO] [1637885593.321712976] [mavros]: Plugin waypoint added to executor
[INFO] [1637885593.321728523] [mavros]: Plugin waypoint initialized
[INFO] [1637885593.333418959] [mavros]: Plugin wind_estimation created
[INFO] [1637885593.333487582] [mavros]: Plugin wind_estimation added to executor
[INFO] [1637885593.333518224] [mavros]: Plugin wind_estimation initialized
[INFO] [1637885593.334596826] [mavros]: Built-in SIMD instructions: SSE, SSE2
[INFO] [1637885593.334616824] [mavros]: Built-in MAVLink package version: 2021.10.10
[INFO] [1637885593.334624863] [mavros]: Known MAVLink dialects: common ardupilotmega ASLUAV all development icarous matrixpilot paparazzi standard uAvionix ualberta
[INFO] [1637885593.334631328] [mavros]: MAVROS UAS via /uas1 started. MY ID 1.191, TARGET ID 1.1
[INFO] [1637885594.290298681] [mavros_router]: link[1002] detected remote address 1.191
[WARN] [1637885594.333568819] [mavros.cmd]: CMD: Unexpected command 520, result 0
[INFO] [1637885594.333970994] [mavros.geofence]: GF: Using MISSION_ITEM_INT
[INFO] [1637885594.334035868] [mavros.rallypoint]: RP: Using MISSION_ITEM_INT
[INFO] [1637885594.334071332] [mavros.mission]: WP: Using MISSION_ITEM_INT
[INFO] [1637885594.334115447] [mavros.sys]: VER: 1.1: Capabilities         0x000000000000fbef
[INFO] [1637885594.334147263] [mavros.sys]: VER: 1.1: Flight software:     040003ff (0032363830646666)
[INFO] [1637885594.334174287] [mavros.sys]: VER: 1.1: Middleware software: 00000000 (0000000000000000)
[INFO] [1637885594.334238730] [mavros.sys]: VER: 1.1: OS software:         00000000 (0000000000000000)
[INFO] [1637885594.334262476] [mavros.sys]: VER: 1.1: Board hardware:      00000000
[INFO] [1637885594.334291335] [mavros.sys]: VER: 1.1: VID/PID:             0000:0000
[INFO] [1637885594.334307393] [mavros.sys]: VER: 1.1: UID:                 0000000000000000
[INFO] [1637885594.409823706] [mavros]: CON: Got HEARTBEAT, connected. FCU: ArduPilot
[INFO] [1637885594.410069179] [mavros.mission]: WP: detected enable_partial_push: 1
[INFO] [1637885595.684316799] [mavros.sys]: FCU: EKF2 IMU0 initial yaw alignment complete
[INFO] [1637885595.684460761] [mavros.sys]: FCU: EKF2 IMU1 initial yaw alignment complete
[INFO] [1637885596.724165067] [mavros.sys]: FCU: EKF2 IMU0 tilt alignment complete
[INFO] [1637885596.724910587] [mavros.sys]: FCU: EKF2 IMU1 tilt alignment complete
[INFO] [1637885604.415758467] [mavros.home_position]: HP: requesting home position
[INFO] [1637885604.531906914] [mavros.sys]: FCU: ArduCopter V4.0.3 (ffd08628)
[INFO] [1637885604.532044000] [mavros.sys]: FCU: efc62b28baf7524b66582ede60048b9a
[INFO] [1637885604.532229751] [mavros.sys]: FCU: Frame: QUAD
[INFO] [1637885604.816980451] [mavros.param]: PR: parameters list received
[INFO] [1637885605.556077514] [mavros.sys]: FCU: EKF2 IMU0 origin set
[INFO] [1637885605.556231370] [mavros.sys]: FCU: EKF2 IMU1 origin set
[INFO] [1637885609.428239303] [mavros.mission]: WP: mission received
[INFO] [1637885614.443099532] [mavros.rallypoint]: RP: mission received
[INFO] [1637885619.440966561] [mavros.geofence]: GF: mission received
```

### Diagnostics

```
header:
  stamp:
    sec: 1637885684
    nanosec: 874840565
  frame_id: ''
status:
- level: "\0"
  name: 'mavros_router: MAVROS Router'
  message: ok
  hardware_id: none
  values:
  - key: Endpoints
    value: '3'
  - key: Messages routed
    value: '1319'
  - key: Messages sent
    value: '2627'
  - key: Messages dropped
    value: '0'
- level: "\0"
  name: 'mavros_router: endpoint 1000: tcp://localhost:5760'
  message: ok
  hardware_id: none
  values:
  - key: Received packets
    value: '1300'
  - key: Dropped packets
    value: '0'
  - key: Buffer overruns
    value: '0'
  - key: Parse errors
    value: '0'
  - key: Rx sequence number
    value: '20'
  - key: Tx sequence number
    value: '0'
  - key: Rx total bytes
    value: '46696'
  - key: Tx total bytes
    value: '436'
  - key: Rx speed
    value: '21.000000'
  - key: Tx speed
    value: '0.000000'
  - key: Remotes count
    value: '3'
  - key: Remote [0]
    value: '0.0'
  - key: Remote [1]
    value: '1.0'
  - key: Remote [2]
    value: '1.1'
- level: "\0"
  name: 'mavros_router: endpoint 1001: udp://@192.168.1.117'
  message: ok
  hardware_id: none
  values:
  - key: Received packets
    value: '0'
  - key: Dropped packets
    value: '0'
  - key: Buffer overruns
    value: '0'
  - key: Parse errors
    value: '0'
  - key: Rx sequence number
    value: '0'
  - key: Tx sequence number
    value: '0'
  - key: Rx total bytes
    value: '0'
  - key: Tx total bytes
    value: '46695'
  - key: Rx speed
    value: '0.000000'
  - key: Tx speed
    value: '21.000000'
  - key: Remotes count
    value: '1'
  - key: Remote [0]
    value: '0.0'
- level: "\0"
  name: 'mavros_router: endpoint 1002: /uas1'
  message: ok
  hardware_id: none
  values:
  - key: Remotes count
    value: '3'
  - key: Remote [0]
    value: '0.0'
  - key: Remote [1]
    value: '1.0'
  - key: Remote [2]
    value: '1.191'
---
header:
  stamp:
    sec: 1637885685
    nanosec: 38278033
  frame_id: ''
status:
- level: "\0"
  name: 'mavros: MAVROS UAS'
  message: connected
  hardware_id: uas:///uas1
  values: []
- level: "\x02"
  name: 'mavros: GPS'
  message: No satellites
  hardware_id: uas:///uas1
  values:
  - key: Satellites visible
    value: '0'
  - key: Fix type
    value: '0'
  - key: EPH (m)
    value: Unknown
  - key: EPV (m)
    value: Unknown
- level: "\0"
  name: 'mavros: System'
  message: Normal
  hardware_id: uas:///uas1
  values:
  - key: Sensor present
    value: '0x00000000'
  - key: Sensor enabled
    value: '0x00000000'
  - key: Sensor health
    value: '0x00000000'
  - key: CPU Load (%)
    value: '0.0'
  - key: Drop rate (%)
    value: '0.0'
  - key: Errors comm
    value: '0'
  - key: 'Errors count #1'
    value: '0'
  - key: 'Errors count #2'
    value: '0'
  - key: 'Errors count #3'
    value: '0'
  - key: 'Errors count #4'
    value: '0'
- level: "\x02"
  name: 'mavros: Battery'
  message: No data
  hardware_id: uas:///uas1
  values:
  - key: Voltage
    value: '-1.00'
  - key: Current
    value: '0.0'
  - key: Remaining
    value: '0.0'
- level: "\0"
  name: 'mavros: Heartbeat'
  message: Normal
  hardware_id: uas:///uas1
  values:
  - key: Heartbeats since startup
    value: '90'
  - key: Frequency (Hz)
    value: '1.000027'
  - key: Vehicle type
    value: Quadrotor
  - key: Autopilot type
    value: ArduPilot
  - key: Mode
    value: STABILIZE
  - key: System status
    value: STANDBY
---
header:
  stamp:
    sec: 1637885685
    nanosec: 873956639
  frame_id: ''
status:
- level: "\0"
  name: 'mavros_router: MAVROS Router'
  message: ok
  hardware_id: none
  values:
  - key: Endpoints
    value: '3'
  - key: Messages routed
    value: '1321'
  - key: Messages sent
    value: '2631'
  - key: Messages dropped
    value: '0'
- level: "\0"
  name: 'mavros_router: endpoint 1000: tcp://localhost:5760'
  message: ok
  hardware_id: none
  values:
  - key: Received packets
    value: '1302'
  - key: Dropped packets
    value: '0'
  - key: Buffer overruns
    value: '0'
  - key: Parse errors
    value: '0'
  - key: Rx sequence number
    value: '22'
  - key: Tx sequence number
    value: '0'
  - key: Rx total bytes
    value: '46754'
  - key: Tx total bytes
    value: '436'
  - key: Rx speed
    value: inf
  - key: Tx speed
    value: -nan
  - key: Remotes count
    value: '3'
  - key: Remote [0]
    value: '0.0'
  - key: Remote [1]
    value: '1.0'
  - key: Remote [2]
    value: '1.1'
- level: "\0"
  name: 'mavros_router: endpoint 1001: udp://@192.168.1.117'
  message: ok
  hardware_id: none
  values:
  - key: Received packets
    value: '0'
  - key: Dropped packets
    value: '0'
  - key: Buffer overruns
    value: '0'
  - key: Parse errors
    value: '0'
  - key: Rx sequence number
    value: '0'
  - key: Tx sequence number
    value: '0'
  - key: Rx total bytes
    value: '0'
  - key: Tx total bytes
    value: '46753'
  - key: Rx speed
    value: -nan
  - key: Tx speed
    value: inf
  - key: Remotes count
    value: '1'
  - key: Remote [0]
    value: '0.0'
- level: "\0"
  name: 'mavros_router: endpoint 1002: /uas1'
  message: ok
  hardware_id: none
  values:
  - key: Remotes count
    value: '3'
  - key: Remote [0]
    value: '0.0'
  - key: Remote [1]
    value: '1.0'
  - key: Remote [2]
    value: '1.191'
---
header:
  stamp:
    sec: 1637885686
    nanosec: 38322943
  frame_id: ''
status:
- level: "\0"
  name: 'mavros: MAVROS UAS'
  message: connected
  hardware_id: uas:///uas1
  values: []
- level: "\x02"
  name: 'mavros: GPS'
  message: No satellites
  hardware_id: uas:///uas1
  values:
  - key: Satellites visible
    value: '0'
  - key: Fix type
    value: '0'
  - key: EPH (m)
    value: Unknown
  - key: EPV (m)
    value: Unknown
- level: "\0"
  name: 'mavros: System'
  message: Normal
  hardware_id: uas:///uas1
  values:
  - key: Sensor present
    value: '0x00000000'
  - key: Sensor enabled
    value: '0x00000000'
  - key: Sensor health
    value: '0x00000000'
  - key: CPU Load (%)
    value: '0.0'
  - key: Drop rate (%)
    value: '0.0'
  - key: Errors comm
    value: '0'
  - key: 'Errors count #1'
    value: '0'
  - key: 'Errors count #2'
    value: '0'
  - key: 'Errors count #3'
    value: '0'
  - key: 'Errors count #4'
    value: '0'
- level: "\x02"
  name: 'mavros: Battery'
  message: No data
  hardware_id: uas:///uas1
  values:
  - key: Voltage
    value: '-1.00'
  - key: Current
    value: '0.0'
  - key: Remaining
    value: '0.0'
- level: "\0"
  name: 'mavros: Heartbeat'
  message: Normal
  hardware_id: uas:///uas1
  values:
  - key: Heartbeats since startup
    value: '90'
  - key: Frequency (Hz)
    value: '0.900363'
  - key: Vehicle type
    value: Quadrotor
  - key: Autopilot type
    value: ArduPilot
  - key: Mode
    value: STABILIZE
  - key: System status
    value: STANDBY
---
```

### Check ID

```
No executable found
```

**Comment by vooon** [comment]
Duplicate of #1588

**Comment by weicong96** [comment]
@vooon 
Fyi, It appears that the above situation is resolved after making set_stream_rate run and wait in the background before starting up the node. I now have telemetry streaming but calling the set_mode service does not have any response that comes back. The set_mode call:
  ```
  response = None
  req = SetMode.Request()
  req.custom_mode = 'GUIDED'

  set_mode_future = self.set_mode_client.call_async(req)
  while rclpy.ok():
    try:
      if set_mode_future.done():
        response = set_mode_future.result()
        break
    except Exception as e:
      self.log.error(str(e))
  return response
```

I also saw that in set_mode message definition([link](http://docs.ros.org/en/api/mavros_msgs/html/srv/SetMode.html)) , there is a base_mode field, i'm wondering if i'm doing something wrong? Also tried with base_mode set to one of the enums, none of them give a different behavior.
